File: code_2cases.R
library(data.table)
library(knitr)

# ---------------------------------------------------------
# Core Estimators (Mapping estimating equations to sequential GLMs)
# ---------------------------------------------------------
est_case1 <- function(df) {
  m <- glm(Y ~ A * W, data = df[S == 1], family = binomial)
  
  df1 <- copy(df)[, A := 1]
  df0 <- copy(df)[, A := 0]
  
  y1_hat <- predict(m, newdata = df1, type = "response")
  y0_hat <- predict(m, newdata = df0, type = "response")
  
  rd_std <- mean(y1_hat) - mean(y0_hat)
  rd_mod <- mean(y1_hat[df$A == 1]) - mean(y0_hat[df$A == 0])
  
  return(c("Standard" = rd_std, "Modified" = rd_mod))
}

est_case2 <- function(df) {
  df_s1 <- df[S == 1]
  df1 <- copy(df)[, A := 1]
  df0 <- copy(df)[, A := 0]
  
  # 1-3. Standard G-computation variants
  m_x <- glm(Y ~ A + X, data = df_s1, family = binomial)
  m_z <- glm(Y ~ A + Z, data = df_s1, family = binomial)
  m_xz <- glm(Y ~ A + X + Z, data = df_s1, family = binomial)
  
  rd_x <- mean(predict(m_x, df1, type="response")) - mean(predict(m_x, df0, type="response"))
  rd_z <- mean(predict(m_z, df1, type="response")) - mean(predict(m_z, df0, type="response"))
  
  y1_inner <- predict(m_xz, df1, type="response")
  y0_inner <- predict(m_xz, df0, type="response")
  rd_xz <- mean(y1_inner) - mean(y0_inner)
  
  # 4. Nested G-computation
  df_temp <- copy(df)
  df_temp[, `:=`(y1_inner = y1_inner, y0_inner = y0_inner)]
  
  # Suppress fractional binomial warnings for outer models
  suppressWarnings({
    m_out1 <- glm(y1_inner ~ A + Z, data = df_temp, family = binomial)
    m_out0 <- glm(y0_inner ~ A + Z, data = df_temp, family = binomial)
  })
  
  rd_nested <- mean(predict(m_out1, df1, type="response")) - mean(predict(m_out0, df0, type="response"))
  
  return(c("X-only" = rd_x, "Z-only" = rd_z, "X+Z" = rd_xz, "Nested" = rd_nested))
}

# ---------------------------------------------------------
# Bootstrap Wrapper & Table Generation
# ---------------------------------------------------------
run_comparison_table <- function(df, est_func, table_name, B=500) {
  n <- nrow(df)
  m <- floor(n * 0.9)
  d <- floor(n * 0.1)
  
  pt_ests <- est_func(df)
  k_names <- names(pt_ests)
  
  np_boots <- matrix(NA, nrow=B, ncol=length(k_names))
  m_boots <- matrix(NA, nrow=B, ncol=length(k_names))
  jk_boots <- matrix(NA, nrow=B, ncol=length(k_names))
  
  set.seed(42)
  cat(sprintf("\nRunning Bootstraps for %s (B=%d)...\n", table_name, B))
  
  for (i in 1:B) {
    np_boots[i, ] <- est_func(df[sample(1:n, n, replace = TRUE)])
    m_boots[i, ]  <- est_func(df[sample(1:n, m, replace = TRUE)])
    jk_boots[i, ] <- est_func(df[sample(1:n, (n-d), replace = FALSE)])
  }
  
  # Compile metrics
  results <- data.frame(Estimator = k_names, Point_Estimate = pt_ests, row.names = NULL)
  
  # Column calculations leveraging scaled variances for accurate CIs
  results$`NP-Boot SE` <- apply(np_boots, 2, sd)
  results$`NP 95% CI` <- sprintf("[%gre, %gre]", 
                                 pt_ests - 1.96*results$`NP-Boot SE`, 
                                 pt_ests + 1.96*results$`NP-Boot SE`)
  
  results$`m-Boot SE` <- apply(m_boots, 2, sd) * sqrt(m/n)
  results$`m-Boot CI` <- sprintf("[%gre, %gre]", 
                                 pt_ests - 1.96*results$`m-Boot SE`, 
                                 pt_ests + 1.96*results$`m-Boot SE`)
  
  results$`Jackknife SE` <- apply(jk_boots, 2, sd) * sqrt((n-d)/d)
  results$`Jackknife CI` <- sprintf("[%gre, %gre]", 
                                    pt_ests - 1.96*results$`Jackknife SE`, 
                                    pt_ests + 1.96*results$`Jackknife SE`)
  
  cat(paste0("\n=== ", table_name, " ===\n"))
  print(kable(results, digits = 3, align = 'l'))
}

# Execution
d1 <- fread("data/example1.csv")
d2 <- fread("data/example2.csv")

run_comparison_table(d1, est_case1, "Table 1: Case 1 Results", B=500)
run_comparison_table(d2, est_case2, "Table 2: Case 2 Results", B=500)

File: code_2cases.py
import numpy as np
import pandas as pd
import statsmodels.formula.api as smf
import statsmodels.api as sm
from tqdm import tqdm
import warnings

warnings.filterwarnings('ignore')

# ---------------------------------------------------------
# Core Estimators (Mapping estimating equations to sequential GLMs)
# ---------------------------------------------------------
def fit_glm(formula, data):
    """Helper to fit logistic regression."""
    return smf.glm(formula, data=data, family=sm.families.Binomial()).fit(disp=False)

def est_case1(df):
    """Table 1: Case 1 Estimators"""
    res = {}
    df_s1 = df[df['S'] == 1]
    
    # 1. Complete-case analysis (naïve)
    y1_naive = df_s1[df_s1['A'] == 1]['Y'].mean()
    y0_naive = df_s1[df_s1['A'] == 0]['Y'].mean()
    res['Complete-case analysis (naïve)'] = y1_naive - y0_naive
    
    # 2 & 3. Standard and Modified g-computation
    m = fit_glm("Y ~ A * W", df_s1)
    df1, df0 = df.assign(A=1), df.assign(A=0)
    
    y1_hat, y0_hat = m.predict(df1), m.predict(df0)
    res['Standard g-computation'] = y1_hat.mean() - y0_hat.mean()
    res['Modified g-computation'] = y1_hat[df['A'] == 1].mean() - y0_hat[df['A'] == 0].mean()
    
    # 4. Inverse probability weighting
    # Selection model based on A and W (treatment induced selection)
    m_s = fit_glm("S ~ A * W", df)
    pi_s = m_s.predict(df)
    ipw = df['S'] / pi_s
    
    # Weighted means
    wt1, wt0 = df['A'] * ipw, (1 - df['A']) * ipw
    res['Inverse probability weighting'] = (wt1 * df['Y']).sum() / wt1.sum() - (wt0 * df['Y']).sum() / wt0.sum()
    
    return res

def est_case2(df):
    """Table 2: Case 2 Estimators"""
    res = {}
    df_s1 = df[df['S'] == 1]
    df1, df0 = df.assign(A=1), df.assign(A=0)
    
    # 1. Complete-case analysis (naïve)
    y1_naive = df_s1[df_s1['A'] == 1]['Y'].mean()
    y0_naive = df_s1[df_s1['A'] == 0]['Y'].mean()
    res['Complete-case analysis (naïve)'] = y1_naive - y0_naive
    
    # 2, 3, 4. Standard G-computation variants
    for name, formula in [('Standard g-computation,{Z}', "Y ~ A + Z"), 
                          ('Standard g-computation,{X}', "Y ~ A + X"), 
                          ('Standard g-computation,{X, Z}', "Y ~ A + X + Z")]:
        m = fit_glm(formula, df_s1)
        res[name] = m.predict(df1).mean() - m.predict(df0).mean()
        
    # 5. Iterated g-computation (Nested)
    m_inner = fit_glm("Y ~ A + X + Z", df_s1)
    df_temp = df.copy()
    df_temp['y1_inner'] = m_inner.predict(df1)
    df_temp['y0_inner'] = m_inner.predict(df0)
    
    m_out1 = fit_glm("y1_inner ~ A + Z", df_temp)
    m_out0 = fit_glm("y0_inner ~ A + Z", df_temp)
    res['Iterated g-computation'] = m_out1.predict(df1).mean() - m_out0.predict(df0).mean()
    
    # 6. Inverse probability weighting
    # Selection depends on X, Treatment depends on Z
    m_s = fit_glm("S ~ X", df)
    m_a = fit_glm("A ~ Z", df)
    
    pi_s = m_s.predict(df)
    pi_a = m_a.predict(df)
    
    # Calculate weights exactly as defined in estfun.py psi_ipw_case2
    ipmw = df['S'] / pi_s
    iptw = df['A'] / pi_a + (1 - df['A']) / (1 - pi_a)
    ipw = ipmw * iptw
    
    wt1, wt0 = df['A'] * ipw, (1 - df['A']) * ipw
    res['Inverse probability weighting'] = (wt1 * df['Y']).sum() / wt1.sum() - (wt0 * df['Y']).sum() / wt0.sum()
    
    return res

# ---------------------------------------------------------
# Bootstrap Wrapper & Table Generation
# ---------------------------------------------------------
def run_comparison_table(df, est_func, table_name, B=500):
    n = len(df)
    m = int(n * 0.9)  # m-out-of-n size
    d = int(n * 0.1)  # delete-d size
    
    pt_ests = est_func(df)
    
    np_boots = {k: [] for k in pt_ests.keys()}
    m_boots = {k: [] for k in pt_ests.keys()}
    jk_boots = {k: [] for k in pt_ests.keys()}
    
    np.random.seed(42)
    print(f"\nRunning Bootstraps for {table_name} (B={B})...")
    for _ in tqdm(range(B)):
        # 1. Non-parametric
        df_np = df.sample(n=n, replace=True)
        for k, v in est_func(df_np).items(): np_boots[k].append(v)
            
        # 2. m-out-of-n
        df_m = df.sample(n=m, replace=True)
        for k, v in est_func(df_m).items(): m_boots[k].append(v)
            
        # 3. Delete-d Jackknife
        df_jk = df.sample(n=n-d, replace=False)
        for k, v in est_func(df_jk).items(): jk_boots[k].append(v)

    # Compile Table Data
    records = []
    for k in pt_ests.keys():
        pt = pt_ests[k]
        np_se = np.std(np_boots[k])
        m_se = np.std(m_boots[k]) * np.sqrt(m/n)
        jk_se = np.std(jk_boots[k]) * np.sqrt((n-d)/d)
        
        records.append({
            'Estimator': k,
            'Point Estimate': pt,
            'NP-Boot SE': np_se,
            'NP 95% CI': f"[{pt - 1.96*np_se:.3f}, {pt + 1.96*np_se:.3f}]",
            'm-Boot SE': m_se,
            'm-Boot CI': f"[{pt - 1.96*m_se:.3f}, {pt + 1.96*m_se:.3f}]",
            'Jackknife SE': jk_se,
            'Jackknife CI': f"[{pt - 1.96*jk_se:.3f}, {pt + 1.96*jk_se:.3f}]"
        })
        
    df_res = pd.DataFrame(records).set_index('Estimator')
    print(f"\n=== {table_name} ===")
    print(df_res.round(3).to_markdown())

# Execution
if __name__ == "__main__":
    d1 = pd.read_csv("data/example1.csv")
    d2 = pd.read_csv("data/example2.csv")
    
    run_comparison_table(d1, est_case1, "Table 1: Case 1 Results", B=500)
    run_comparison_table(d2, est_case2, "Table 2: Case 2 Results", B=500)
